### PR TITLE
Add SegmentProdGrad

### LINKED ISSUE #41090: Issue using segment_prod in custom keras layer
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): colab
- TensorFlow installed from (source or binary): nightly
- TensorFlow version (use command below): nightly
- Python version: colab

I'm trying to write a custom keras layer that uses `segment_prod` on model features, i.e., not the batch dimension. To do that, I've been using tf.transpose to put the feature dimensions first, and then transposing the resulting calculation. Forward calculation seems to work using this approach, but there appears to be an issue with gradients with `segment_prod`: the following code fails with `LookupError: gradient registry has no entry for: SegmentProd`:

```python
class MyLayer1(layers.Layer):

  def call(self, inputs):

    segments = tf.constant([0, 0, 0, 1, 1])
    return tf.transpose(
        tf.math.segment_prod(tf.transpose(inputs), segments))

inputs = layers.Input(10)
embed = layers.Embedding(20, 5)(inputs)
output = MyLayer1()(embed)
output = layers.GlobalAveragePooling1D()(output)
model = tf.keras.Model(inputs, output)

X = np.random.randint(20, size=(100, 10))
y = np.random.randn(100,2)

model.compile(loss='mae')
model.fit(X, y)
```

Replacing `tf.math.segment_prod` with other operations, like `tf.math.unsorted_segment_prod(..., num_segments=2)` or `tf.math.segment_sum` seems to work, however.

Colab gist:
https://colab.research.google.com/gist/pstjohn/8fa2faf274679742ba628290a94c2b2b/untitled0.ipynb

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes https://github.com/tensorflow/tensorflow/issues/41090.

### CODE REVIEW

File: tensorflow/python/ops/math_grad.py [hunk 1]
```diff
@@ -414,6 +414,48 @@ def _SegmentMaxGrad(op, grad):
...
+  grad = array_ops.where_v2(
+      math_ops.greater(num_zeros, 1), array_ops.zeros_like(grad), grad)
+  # replace all zeros with ones and compute the segment_prod
+  non_zero_data = array_ops.where_v2(is_zero,
+                                     array_ops.ones_like(data),
+                                     data)
+  non_zero_prod = gen_math_ops.segment_prod(non_zero_data, segment_ids)
+  gathered_prod = array_ops.gather(op.outputs[0], segment_ids)
+  gathered_non_zero_prod = array_ops.gather(non_zero_prod, segment_ids)
+  prod_divided_by_el = gathered_prod / data  # May contain nan/inf.
```
Comment (alextp) [inline tensorflow/python/ops/math_grad.py:450]: Shouldn't we use non_zero_data here? Or only divide gathered_prod by data when either gathered_prod or data is nonzero?

Comment (WindQAQ) [reply tensorflow/python/ops/math_grad.py:450]: Good point! I think it doesn't really matter as we have use `where` to select the finite parts, but the change indeed makes the codes clearer. I'll update accordingly. Thank you!

**Comment by gbaned** [comment]
@rohan100jain Can you please take a look on this PR ? Thanks!

**Review by rohan100jain — COMMENTED** [review]
Looks fine to me. Saurabh can you also take a look?

**Review by saxenasaurabh — APPROVED** [review]
Very nice!

**Pull request merged**